include/zephyr/drivers/flash.h

```diff
@@ -125,6 +125,8 @@ typedef void (*flash_api_pages_layout)(const struct device *dev,
 typedef int (*flash_api_sfdp_read)(const struct device *dev, off_t offset,
 				   void *data, size_t len);
 typedef int (*flash_api_read_jedec_id)(const struct device *dev, uint8_t *id);
+typedef int (*flash_api_ex_op)(const struct device *dev, uint16_t code,
+			       uintptr_t *in, uintptr_t *out);
```

Comment:
I am not yet convinced about the function parameters. Would it be clearer to pass instead:

 `device *dev, uint32_t code, const void  *code_param`

Reply:
I'm fine with one pointer instead of two. One note: we will need to have separate codes to `get` and `set` option (which is also fine). I would remove also `const` (so, `const void  *code_param` will become `void *code_data`), because we need to pass data from `get` codes to the user.

Do you strongly prefer `void *` over `uintptr_t *`?

I will wait for comments from @de-nordic before changing this.

Reply:
> I'm fine with one pointer instead of two. One note: we will need to have separate codes to `get` and `set` option (which is also fine). I would remove also `const` (so, `const void *code_param` will become `void *code_data`), because we need to pass data from `get` codes to the user.
>

I would pass the code_data pointer as a `const void *`, if it would be a "get" operation it would contain a pointer to where the data can be written (not a const) and a size that has been allocated for that. In a "set" operation it is also possible to pass the size. So the idea was not to use this directly but only after a cast to something containing a size and a buffer location.
 
> Do you strongly prefer `void *` over `uintptr_t *`?
>

`void *` is used more than `uintptr_t *` in the code and make it more clear that it is expected to be casted.

> I will wait for comments from @de-nordic before changing this.

OK

Reply:
It is more general and versatile, suppose that for some operation you would need 2 buffers each with their own size, this wouldn't fit in the pointer + size setup. As a bonus this also only needs 1 extra api function and not 2.

Maybe it would be better to use "context" (`op_ctx`) instead of `code_param`.

Reply:
To further elaborate this: in my view this command should support everything, even reads and writes should be mappable to a `flash_ex_op` call, as well as reads to ecc bits for nand flash, reads and writes to special registers, enable power down, ...

Reply:
I understand the https://github.com/zephyrproject-rtos/zephyr/pull/53441#discussion_r1095369131.

and try to balance 

> I'm not sure if this is better than passing pointer and size in a `flash_ex_op` call. From the user perspective, in order to get a write protect status I need to allocate two structures: The first for a pointer and size, the second for the results. For me, this sounds like an overkill, but maybe I'm missing something.

When you would be calling five param syscall, all of these params need to be passed to syscall, whether they are used or not (so set to NULL, for example). This means that compiler first needs to store the contents of registers it is going to use to pass parameters before passing current parameters and doing the call.
So you are going to use, implicitly, some extra stack for each call and extra code to take care about saving registers, passing parameters and restoring registers after call. I understand that.

But I will side with @semihalf-duda-patryk here because I find it easier and more convenient to allocate space, either on stack or in bss, for in/out structures separately and just pass them when needed to syscall; note that not all functions may require in object, not all may require out object, some may require none, but in case when either in or out is needed, the intermediate context object will also need to be provided.
Additionally it allows us to set `in` pointer as const, which means that we should be able to directly take pointer to flash and pass it to syscall, with no need to move it to RAM first (either stack or bss allocated thing).

The `const uintptr_t in` (should be `uintptr_t`, not `uintptr_t *`, which was my mistake) was proposed because, I was thinking about direct passing integer parameters via registers, for example when you have some function like:
```
flash_ex_op(dev, SET_SLEEP_LEVEL, x, NULL) // x is level in range 0-N
flash_ex_op(dev, GET_POWER_INFO_BLOCK, x, p) // where x is info index that we want, and p is x specific type pointer to structure where the info will be placed
```
to avoid needing to do `copyin` of  whatever `in` points, to kernel space, before accessing it, when it is not needed - for example to process single int param.
`out` should be `void *`, `intptr_t` does not make sense here);

Also having separate `out`  allows syscalls to copyout data only on demand, and allows to easily discard return object without doing copyout of it; for example you may have syscall that would return some object, as a result of operation, that you only need in certain situations, so in that situations you pass the `out` pointer as non-NULL and get the data; in all other instances you just pass NULL and syscall will not bother spending time copying something you will not look into anyway.

> To further elaborate this: in my view this command should support everything, even reads and writes should be mappable to a `flash_ex_op` call, as well as reads to ecc bits for nand flash, reads and writes to special registers, enable power down, ...

Yes it will, it will only require defining operation codes, expected format of in and/or out objects and return codes.

Reply:
I agree that setting up a "context" before passing it to the function is a nuisance and should be avoided if possible. However passing a pointer without a length as an output argument should not be allowed as this is prone to errors and might lead to overwritten stack data, ... 

Adding a length is an option but this would increase the parameters to the function (and would probably be needed also for the input ptr e.g. when performing a write as a `flash_ex_op` call).

When a input and output structure would need to be created anyhow (to pass the allocated buffer sizes) it is a small step to integrate them together and to pass only one structure (that would contain pointers to buffers and the allocated buffer size). Depending on the operation (wether an output is needed or not) this could be casted to the appropiate structure.

This function should be as general purpose as possible so I think it would be good if we already have an idea how existing functions would be mapped to the `flash_ex_op` call, the read/write/erase, flash_layout and spi_nor_flash specific commands. When it is possible to map all of these we are on track of having selected a good api extension.

Reply:
> I agree that setting up a "context" before passing it to the function is a nuisance and should be avoided if possible. However passing a pointer without a length as an output argument should not be allowed as this is prone to errors and might lead to overwritten stack data, ...

Yes and no. Generally we can specify what does function return and limit it to certain objects and expect only space for such objects. We may have problems here with pre-compiled object with different definitions of expected returned objects, but this will trigger need for recompile anyway as soon as the problem is detected.

> 
> Adding a length is an option but this would increase the parameters to the function (and would probably be needed also for the input ptr e.g. when performing a write as a `flash_ex_op` call).

Input ptr will not require it because we know what we are passing when the pointer reaches the syscall. As an example from FreeBSD:
https://svnweb.freebsd.org/base/stable/12/sys/kern/kern_fork.c?view=markup#l148
https://svnweb.freebsd.org/base/stable/12/sys/kern/kern_exit.c?view=markup#l172
both syscalls get already pointer of specified type they expect here.
Having size could help us with some variable length data output or enumeration, but that would work until the size changes or number of enumerated items is to big, which basically means changes to buffer sizes in code and recompilation.

> 
> When a input and output structure would need to be created anyhow (to pass the allocated buffer sizes) it is a small step to integrate them together and to pass only one structure (that would contain pointers to buffers and the allocated buffer size). Depending on the operation (wether an output is needed or not) this could be casted to the appropiate structure.
> 

Yest, but this just brings extra gymnastics do something and does not add that much flexibility.

> This function should be as general purpose as possible so I think it would be good if we already have an idea how existing functions would be mapped to the `flash_ex_op` call, the read/write/erase, flash_layout and spi_nor_flash specific commands. When it is possible to map all of these we are on track of having selected a good api extension.

Yes. So we will reserve the first IDs for the read, write, erase, and all other callbacks that have already been provided by the Flash API, but this is not that urgent since, at this point, we have established where the barrier between system defined and platform/vendor  specific codes are, and while it will be quite easy to add vendor codes, addition of new system codes would require some extra evaluation.
Reserving codes for know Flash API calls would be only formal thing, for now, because I do not think that we should have flash_ex_op defined for them, unless the flash_ex_op provides some extra functionality the default flash API call does not provide.
Also we may end up adding direct Flash API calls for functionalities that get called often, for example power save functions may get called frequently enough that calling them via `flash_ex_op`  may be waste of clocks.

Reply:
Ok, is it a good idea to create the difference between read and write constructs ? It will almost certain be needed to create a struct when doing a "get" operation to pass the allocated buffer size on top of some extra info (e.g. an offset), so why not pass them all as a struct (like it is done for the write.

Reply:
```
struct ex_op_rw_ctx {
    uint32_t offset;
    uint32_t size;
    uint8_t *buf;
};


const struct ex_op_rw_ctx  read_here = { 0, sizeof(out_buf), out_buf };

...
ret = flash_ex_op(dev, READ, &read_here);
...
ret = flash_ex_op(dev, SET_POWER_LEVEL, xx);
...
```
Does not make that much difference, but note that this, in case off RAD, re-implemets functions that we have already direct API calls for and does not seem that we are going to have many more of the off-size-buffer functions.
For single in parameter functions, like the SET_POWER_LEVEL above, it makes no difference.

In case of set-get type calls you have:
```
struct in_out_flags = {
    uint32_t in_flags;
    uint32_t *out_flags;
} = { 0xff, NULL }; // not wanting anything back, this time
ret = flash_ex_op(dev, SET_SOME_FLAGS, &in_out_flags);
```
vs
```
uint32_t flags = 0xff;
ret = flash_ex_op(dev, SET_SOME_FLAGS, &flags, NULL);
//or
ret = flash_ex_op(dev, SET_SOME_FLAGS, 0xff, NULL);
//or when you want old flags back:
ret = flash_ex_op(dev, SET_SOME_FLAGS, 0xff, &flags);
```

Lets get page info by offset:
```

// Used to get page info, sector info, protection range, etc.
struct block_info {
    uint32_t offset;
    uint32_t size;
};
struct get_page_info {
    uint32_t offset;
    struct block_info *out;
};

struct block_info bi;
const struct get_page_info gpi = { 0, &bi };
ret = flash_ex_op(dev, GET_PAGE_INFO, &gpi);
...
```
vs

```
struct block_info bi;
uint32_t offset = 0;
ret = flash_ex_op(dev, GET_PAGE_INFO, &offset, &bi);
```

Read 4k page by offset, assumes that buffer is allocated by size of page:
```
struct read_block {
    uint32_t offset;
    uint8_t *out;
} rb = { 0, buf };
ret = flash_ex_op(dev, READ_4K_PAGE, &rb);
```
vs
```
uint32_t offset = 0;
ret = flash_ex_op(dev, READ_4K_PAGE, &offset, buf);
```

Erase page by offset is an example one in-only call:
```
uint32_t offset;
ret = flash_ex_op(dev, ERASE_4K_PAGE, &offset);
```
vs
```
uint32_t offset;
ret = flash_ex_op(dev, ERASE_4K_PAGE, &offset, NULL);
```

Reply:
I agree that the difference is small, but I don't see a reason to provide the *out. IMHO this is better replaced by a sizeof the struct that is passed to avoid casting errors/reading outside struct boundaries.

Probably we would also need to define some "standard" struct layouts (in flash.h) to be passed to this function:

1. Structure for a simple command passed,
2. Structure for a read type command passed,
3. Structure for a write type command passed,
4. Structure for a read+write type command passed (with 2 buffers),

As an example of the usage of the last I'm thinking of flash devices that are combined with other peripherals where a specific write command needs to be passed to get the desired output.

Reply:
> I agree that the difference is small, but I don't see a reason to provide the *out. IMHO this is better replaced by a sizeof the struct that is passed to avoid casting errors/reading outside struct boundaries.

Size does not matter. The driver should only understand predefined structures as op context, there is no point to size them. If we provide size, how do we use it?
Do we compare it against known sizes and drop unsupported? Does not seem to practical, it is rather one time error.
Do we allow any size?
What if user mismatches size between structures? Does giving size saves us from anything? We will just overwrite or read pass what we should, and this is in system mode, so we will not figure that out prior to returning to user mode anyway.


> 
> Probably we would also need to define some "standard" struct layouts (in flash.h) to be passed to this function:
> 
>     1. Structure for a simple command passed,
> 
>     2. Structure for a read type command passed,
> 
>     3. Structure for a write type command passed,
> 
>     4. Structure for a read+write type command passed (with 2 buffers),
> 
> 

We already have API call for most of above cases so we know how they behave, and they are unique themselves and they already do not cover what we are trying to do, they cover what we had already done; the purpose of this API call, and the vendor range itself, is so that we could evaluate things we do not know yet by observing how users try to overcome problems in real scenarios. Then when we collect enough data, or what  kind of problems users have how do they prefer to use API, how certain functionalities are provided by different vendors and then design common op code, for these functionalities, in system code space.

> As an example of the usage of the last I'm thinking of flash devices that are combined with other peripherals where a specific write command needs to be passed to get the desired output.

If these peripherals are not flash related then there is no point to cover them under flash API.

Reply:
> 
> You might have overlooked the "type" that I inserted before "command ..." (and forgotten for the "simple" case 😀 ): a. the get protection would be a "read type command structure", b. the set protection would be a "write type command structure",

I might have. 

> 
> If these structures are not defined how would they be included by the user ? Would users have to include a vendor specific include file ? Where should these include files be placed ?

Yes. That is the idea  https://github.com/zephyrproject-rtos/zephyr/issues/53905#issuecomment-1406687637. When we standardize feature with system wide code, than user can switch to using it.
Non standard codes would not likely to be used in Zephyr code base, as they would start favoring some selected platform.
In case of the RD/WR locks on flash this is actually thing that user will know what they are doing and usage would be app specific, rather then something done in system; but when we finally have some common code and structures derived, after gaining experience with various platforms and mem chips, we could try to evaluate using this to set some app protections, by for example MCUboot, to prevent accidental writes over running application.

I am assemblysaurus from the argsonstack coding period, so passing unused params on stack before doing syscall/int is stuff I can still accept. I remember, though, one of my teachers mentioning to avoid designing more than five param functions to avoid unneeded stack ops, and because x86 didn't have that many registers back then.
Anyway gathering pros and cons here.

Pros/cons for the `flash_ex_op(..., code, const in, out)`

  1.  \+ No need for copy in `in` when only `out` pointer is needed;
  2.  \+ in can be used to pass integer, for simple operations that do not return anything (like set power level);
  3.  \- Need to save-pass-restore register used by `out` and set it to NULL when not used;
  4. \+ `out` can be used to either pass pointer to for object of predefined size, or pointer to structure that will define output and, being non-const, could be filled in in place by function;

Pros/cons for the `flash_ex_op(..., in_out)`

   1. \- No need to copy in `in_out`` to get `out` pointer, unless we agree to make `in_out` non-const, which I would rather not do (vs +on 1  in above list);
   2. same as 2 in above;
   3. \+ No need to save-pass-restore `out` register (vs - on 3 in above list, flash saved)
   4. \- if `in_out` is const and can not be written directly, just for being inconvenient, otherwise the `in_out` may point to some additional structure;

To explain more what I mean by 4:
```
struct something_in {
    uint32_t in_size;
    uint8_t *in_buf;
    uint32_t out_offset;  // Offset where to write not within out buffer;
};
struct something_out {
   uint32_t out_size;
   uint8_t *out_buf;
};
struct something_in sin = { .. some init };
struct something_out sout = { .out_size = max_out_buf_size, .out_buf = out_buf_ptr };
// When successful, the &sout will have .out_size replaced with what size of data written to out_buf_ptr; the function can not directly modify the sin but can do that with sout;
ret = flash_ex_op(dev, code, &sin, &sout);
// Now we have function that does not use extra input info, and can still read/write the out.
ret = flash_ex_op(dev, code, NULL, &sout);
```

For me using separate `in` and `out` is more convenient, at the cost of passing NULL sometimes and one more set of store-pass-restore operations at call. It also allows to easily use the `in` (since the `intptr_t`) as basic integer parameter, for which you do not have to define extra input structure.

Reply:
I would still rather settle on `flash_ex_op(struct device *device, uint32_t code, const void *code_ctx)` because:

a. It is similar to disk `io_ctl`,
b. In the case a size passed for *out, there would be a deviation from first passing the pointer and then the length,
c. Even in the case a size is passed for *out probably another parameter e.g. a offset will need to be passed as well, so there would still be the need to setup a context,
d. The special use case where it is to be expected that only one value will be passed for power level will probably need to be standardized into a new (generic) api function,

Reply:
To make sure that I'm on the same page: by "copy in" and "copy out" you mean accordingly `z_user_from_copy()` and `z_user_to_copy()` used in a verifier as described in https://docs.zephyrproject.org/latest/kernel/usermode/syscalls.html#verification-memory-access-policies? Copy function requires size of structure it's going to copy, but this depends on the extended operation code, so the verifier needs to be implemented in the driver (similar to `subsys/net/ip/net_if.c`) instead of some dedicated file, is this correct?

The link above mentions that we need to make a copy to protect against situation in which userspace changes structure passed to kernel, after it was verified. Does it mean that we don't need to copy in structures that are `const`?

> No need to copy in `in_out` to get `out` pointer, unless we agree to make `in_out` non-const, which I would rather not do (vs +on 1 in above list);

I'm not sure if I understand this correctly. In this case we actually need to copy in in order to get `out` pointer, don't we? Or we can merge input and output parameters in one structure (but it needs to be non-const).

> It is similar to disk `io_ctl`,

`flash_ex_op(struct device *device, uint32_t code, const void *code_ctx)`
`mmc_ioctl(struct sd_card *card, uint8_t cmd, void *buf)`

The difference is that `ioctl` doesn't require the buffer to be `const`. This one keyword affects implementation, so IMO they are not similar.

Reply:
It seems that the `syscall` interface will require the standardization of the "context" passed to the `flash_ex_op` call. Regarding the difference with `mmc_ioctl` call, it is indeed different, but I don't think the `mmc_ioctl` call can be made in accordance with the `syscall` interface (regardless of wether it is a const or not it should pass a size).

I am thinking about a generic way that would allow passing the required info, but I haven't found one yet (even when splitting it up in a `get` and `set` method) taking into account that the parameter count should not be more than 4.